### PR TITLE
c_libax net patch

### PR DESCRIPTION
**Pull request by reflyable**
实现c_libax中网络接口

socket
shutdown
bind
connect
listen
accept
send
recv
sendto
recvfrom
htonl
htons
ntohl
ntohs
inet_pton
inet_ntop
getaddrinfo(功能不完全)
freeaddrinfo

### CODE REVIEW

File: ulib/libax/src/cbindings/file.rs [hunk 1]
```diff
@@ -1,40 +1,12 @@
...
-    for fd in 3..FILE_LIMIT {
-        if fd_table[fd].is_none() {
-            fd_table[fd] = Some(Arc::new(Mutex::new(file)));
-            return Some(fd);
-        }
-    }
-    None
-}
+use axerrno::{LinuxError, LinuxResult};
+use axio::{prelude::*, SeekFrom};
```
Comment (equation314) [inline ulib/libax/src/cbindings/file.rs]: Merge into `use crate::io::{self, ...}`

File: ulib/libax/src/cbindings/fd_table.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        }
+    }
+
+    #[cfg(feature = "fs")]
+    #[allow(irrefutable_let_patterns)]
+    pub fn into_file(self) -> LinuxResult<Arc<Mutex<File>>> {
+        if let Filelike::FileType(s) = self {
+            Ok(s)
+        } else {
+            Err(LinuxError::EPIPE)
```
Comment (equation314) [inline ulib/libax/src/cbindings/fd_table.rs]: `ESPIPE`

File: ulib/libax/src/cbindings/socket.rs [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+        socket.lock().shutdown()?;
+        Ok(0)
+    })
+}
+
+/// Query addresses for a domain name.
+///
+/// Return address number if success.
+#[no_mangle]
+pub unsafe extern "C" fn ax_dnsquery(
```
Comment (equation314) [inline ulib/libax/src/cbindings/socket.rs]: Rename to `ax_resolve_socket_addr`?

File: ulib/libax/src/cbindings/socket.rs [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+        } as *const _ as *const ctypes::sockaddr)
+    };
+}
+
+fn from_c_sockaddr(
+    addr: *const ctypes::sockaddr,
+    addrlen: ctypes::socklen_t,
+) -> LinuxResult<SocketAddr> {
+    if addrlen != size_of::<ctypes::sockaddr>() as u32 || addr.is_null() {
+        return Err(LinuxError::EINVAL);
```
Comment (equation314) [inline ulib/libax/src/cbindings/socket.rs:116]: All null pointers should return `EFAULT`, and all wrong `addrlen` return `EINVAL`.

File: ulib/libax/src/net/udp.rs
Comment (equation314) [inline ulib/libax/src/net/udp.rs:1]: Line 2: `use axnet::{self, SocketAddr}`;

File: ulib/c_libax/src/socket.c [hunk 5]
```diff
@@ -0,0 +1,63 @@
+#include <libax.h>
+#include <stdio.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#if defined(AX_CONFIG_NET)
```
Comment (equation314) [inline ulib/c_libax/src/socket.c:6]: Add new line before

File: ulib/c_libax/src/socket.c [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+int socket(int domain, int type, int protocol)
+{
+    return ax_socket(domain, type, protocol);
+};
```
Comment (equation314) [inline ulib/c_libax/src/socket.c]: Remove `;`, same below.

File: ulib/c_libax/include/netinet/in.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
+#ifndef	_NETINET_IN_H
+#define	_NETINET_IN_H
+#include <sys/socket.h>
```
Comment (equation314) [inline ulib/c_libax/include/netinet/in.h:4]: Add a new line before

**Review by equation314 — APPROVED** [review]

**Review by equation314 — APPROVED** [review]
LGTM. Thanks!

File: ulib/c_libax/src/network.c [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+    if (res_len < 0)
+        return EAI_FAIL;
+    if (res_len == 0)
+        return EAI_NONAME;
+    struct addrinfo *_res = (struct addrinfo *)calloc(res_len, sizeof(struct addrinfo));
+    for (int i = 0; i < res_len; i++) {
+        (_res + i)->ai_family = AF_INET;
+        (_res + i)->ai_addrlen = sizeof(struct sockaddr);
+        (_res + i)->ai_addr = (addrs + i);
+        (_res + i)->ai_next = (_res + i + 1);
```
Comment (coolyjg) [inline ulib/c_libax/src/network.c:30]: This causes bug in Redis:
`ai_socktype` and `ai_protocol` fields are not set. Do not support code like this: 
```
    if ((rv = getaddrinfo(bindaddr,_port,&hints,&servinfo)) != 0) {
        anetSetError(err, "%s", gai_strerror(rv));
        return ANET_ERR;
    }
    for (p = servinfo; p != NULL; p = p->ai_next) {
        if ((s = socket(p->ai_family,p->ai_socktype,p->ai_protocol)) == -1)
            continue;
        .........}
```

**Pull request merged**